### ISSUE TITLE
Speech: Unavailable responses not auto-retried

### ISSUE THREAD

**Issue by breznik** [googleapis/google-cloud-go#1396]
## Client

Speech

## Expected Behavior

Unavailable response codes from Speech service are automatically retried.

## Actual Behavior

It does not appear that StreamingRecognize requests to Cloud Speech which result in an Unavailable response code are automatically retried.

I came across https://godoc.org/github.com/googleapis/gax-go

The retry functionality there seems like its intended more for the implementors of google-cloud-go than for users of google-cloud-go. When I briefly looked at https://github.com/googleapis/google-cloud-go/blob/master/speech/apiv1/speech_client.go#L49 - I see that it is part of the defaultCallOptions but perhaps that is insufficient for it to be actually used? It seems like some other clients (pubsub, spanner, ..) have very specific retry logic. It would be great if cloud speech could have the save behavior built in.

**Comment by jadekler** [comment]
UNAVAILABLE is not always safe to retry. If you would like this behavior, though, you can provide [`gax.WithRetryer` call option](https://godoc.org/github.com/googleapis/gax-go#WithRetry) with a gax retryer that suits your needs (ex [`gax.OnCodes(codes.UNAVAILABLE)`](https://godoc.org/github.com/googleapis/gax-go#OnCodes)).

**Comment by breznik** [comment]
Could you perhaps explain why it's not always safe to retry on Unavailable? I ask because, other parts of google are telling me, that when I encounter Unavailable for cloud speech, I should retry.

**Comment by jadekler** [comment]
The TLDR is that gRPC doesn't guarantee that the data didn't get to the server in the case of UNAVAILABLE. That can be safe for some methods and unsafe for other methods. To solve this problem generally, we are:

- In the process of writing a lot of documentation about how UNAVAILABLE works and how to appropriately retry (will probably be posted at grpc.io) - sorry I don't have that yet!
- Giving backend teams the ability to specify retry config per-RPC, so that if they _know_ a method is safe to retry (usually this is just about the idempotency of a method) they can specify it and cause client libraries to auto-retry
- Giving users the ability to specify retry config (as I did in https://github.com/googleapis/google-cloud-go/issues/1396#issuecomment-484544151 - thankfully speech already has hooks for this)

**Comment by breznik** [comment]
Ok, thanks for the explanation. I understand your original statement was more of a broad blanket statement around Unavailable responses from gRPC services, and not specific to cloud speech.

I am still confused as to the speech library, since the internals looks like it specifies which methods are idempotent ([StreamingReconize](https://github.com/googleapis/google-cloud-go/blob/master/speech/apiv1/speech_client.go#L67) is declared idempotent), and provides the [default](https://github.com/googleapis/google-cloud-go/blob/master/speech/apiv1/speech_client.go#L103) call options to use a retryer, but it's not clear that the retryer is actually working/retrying. Since a retryer is already provided as a call option, it's not clear how if I provide the exact same call option, if anything will change in functionality as you suggested in your [comment](https://github.com/googleapis/google-cloud-go/issues/1396#issuecomment-484544151).

So yes, started this issue as feature request but perhaps it should be bug or question.

**Comment by jadekler** [comment]
Ah! Streaming methods are intentionally not retried. There are additional gotchas for these types of calls. Hope to have more information soon - drafting the article about all this even as we speak (type?), and our internal investigation around standardizing these continues.
